test/test_builders.go

```diff
@@ -0,0 +1,301 @@
+///usr/bin/true; exec /usr/bin/env go run "$0" "$@"
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+)
+
+
+const repo = "ghcr.io/boson-project"
+
+func main() {
+	ctx, cancel := context.WithCancel(context.Background())
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	go func() {
+		<-sigs
+		cancel()
+		<-sigs
+		os.Exit(1)
+	}()
+
+	version := os.Args[1]
+	err := runTests(ctx, version)
+	if err != nil {
+		fmt.Fprintf(os.Stdout, "::error::%s", err.Error())
+		os.Exit(1)
+	}
+}
+
+var builders = []struct {
+	BuilderImage string
+	Runtimes     []string
+}{
+	{
+		BuilderImage: "quay.io/boson/faas-go-builder",
+		Runtimes:     []string{"go"},
+	},
+	{
+		BuilderImage: "quay.io/boson/faas-python-builder",
+		Runtimes:     []string{"python"},
+	},
+	{
+		BuilderImage: "quay.io/boson/faas-nodejs-builder",
+		Runtimes:     []string{"node", "typescript"},
+	},
+	{
+		BuilderImage: "quay.io/boson/faas-jvm-builder",
+		Runtimes:     []string{"quarkus", "springboot"},
+	},
+	{
+		BuilderImage: "quay.io/boson/faas-quarkus-native-builder",
+		Runtimes:     []string{"quarkus"},
+	},
+	{
+		BuilderImage: "quay.io/boson/faas-rust-builder",
+		Runtimes:     []string{"rust"},
+	},
+}
+
+func runTests(ctx context.Context, version string) error {
+
+	os.Setenv("FUNC_REGISTRY", repo)
+
+	for i := range builders {
+		builders[i].BuilderImage = fmt.Sprintf("%s:%s", builders[i].BuilderImage, version)
+	}
+
+	lifecycleVersion := make(map[string]bool)
+
+	for _, builder := range builders {
+		v, err := getLifecycleVersion(builder.BuilderImage)
+		if err != nil {
+			return err
+		}
+		if _, ok := lifecycleVersion[v]; !ok {
+			cmd := exec.CommandContext(ctx, "docker", "pull", "buildpacksio/lifecycle:" + v)
```

Comment:
I like these changes. You mentioned once that we may want to mirror lifecycle images in quay.io. For now just pulling from Docker seems OK to you?

Reply:
@lance fixed